firedrake/functionspaceimpl.py

```diff
@@ -792,6 +800,87 @@ def collapse(self):
         return FunctionSpace(self.mesh(), self.ufl_element())
 
 
+class RestrictedFunctionSpace(FunctionSpace):
+    r"""A representation of a function space, with additional information
+    about where boundary conditions are to be applied.
+
+    If a :class:`FunctionSpace` is represented as V, we can decompose V into
+    V = V0 + VΓ, where V0 contains functions in the basis of V that vanish on
+    the boundary where a boundary condition is applied, and VΓ contains all
+    other basis functions. This means that the :class:`RestrictedFunctionSpace`
+    corresponding to V is V0.
```

Comment:
Modify this to explain that when solving, test and trial functions are drawn from V0 only, and values of the solution in VΓ are treated as constant.